### PR TITLE
Add support for tail calls

### PR DESCRIPTION
**Pull request by antoniofrighetto**
Introduce support for tail calls: any subsequent access in the caller is UB after a tail call-site (bool only approach).

### CODE REVIEW

**Comment by antoniofrighetto** [comment]
There are a few tests (including [this one](https://github.com/AliveToolkit/alive2/blob/master/tests/alive-tv/calls/fn-ptr-taken.srctgt.ll)) which are currently failing because the target function has undergone TailCallElim (as part of optimize_module), thus the refined calls get tail'd. Should we bail out from `check_tailcall` if the call in the source function is not a tail call?

File: tests/alive-tv/bugs/pr43880.srctgt.ll [hunk 1]
```diff
@@ -5,7 +5,7 @@ target triple = "x86_64-unknown-unknown"
 declare i32 @memcmp(ptr nocapture, ptr nocapture, i64)
 
 define i32 @src(ptr nocapture readonly %x, ptr nocapture readonly %y)  {
-  %call = tail call i32 @memcmp(ptr %x, ptr %y, i64 2)
```
Comment (nunoplopes) [inline tests/alive-tv/bugs/pr43880.srctgt.ll:8]: revert this change; it should be fine?

File: tests/alive-tv/calls/fncall-vararg.srctgt.ll [hunk 2]
```diff
@@ -4,11 +4,11 @@ target triple = "x86_64-unknown-linux-gnu"
...
 
 define void @src() {
-  tail call void (ptr, ...) @g(ptr null, i32 0)
+  call void (ptr, ...) @g(ptr null, i32 0)
   ret void
 }
 
 define void @tgt() {
-  tail call void (ptr, ...) @g(ptr null)
+  call void (ptr, ...) @g(ptr null)
```
Comment (nunoplopes) [inline tests/alive-tv/calls/fncall-vararg.srctgt.ll:12]: revert

File: tests/alive-tv/bugs/pr36437-bounded.srctgt.ll [hunk 3]
```diff
@@ -24,7 +24,7 @@ for.body:                                         ; preds = %entry, %for.body
 
 for.end:                                          ; preds = %for.body
   %3 = load i32, ptr @b, align 4
-  %call = tail call i32 (ptr, ...) @print(ptr nonnull dereferenceable(1) getelementptr inbounds ([15 x i8], ptr @.str, i64 0, i64 0), i32 %3)
+  %call = call i32 (ptr, ...) @print(ptr nonnull dereferenceable(1) getelementptr inbounds ([15 x i8], ptr @.str, i64 0, i64 0), i32 %3)
```
Comment (nunoplopes) [inline tests/alive-tv/bugs/pr36437-bounded.srctgt.ll:27]: revert

File: tests/alive-tv/bugs/pr18223.srctgt.ll [hunk 4]
```diff
@@ -74,7 +74,7 @@ for.end:                                          ; preds = %for.end_us_lcssa, %
   store i32 1, ptr @d, align 4
   store i32 %inc.lcssa, ptr @b, align 4
   store i32 %conv.lcssa, ptr @f, align 4
-  %call = tail call i32 (ptr, ...) @myprintf(ptr getelementptr inbounds ([4 x i8], ptr @.str, i64 0, i64 0), i32 %conv.lcssa)
+  %call = call i32 (ptr, ...) @myprintf(ptr getelementptr inbounds ([4 x i8], ptr @.str, i64 0, i64 0), i32 %conv.lcssa)
```
Comment (nunoplopes) [inline tests/alive-tv/bugs/pr18223.srctgt.ll:77]: revert

File: tests/alive-tv/bugs/pr22727.srctgt.ll [hunk 5]
```diff
@@ -39,7 +39,7 @@ if.end:                                           ; preds = %delete.end, %entry
 define i32 @tgt(ptr %bp) {
 entry:
   %add.ptr = getelementptr inbounds i32, ptr %bp, i64 1
-  tail call void @_ZdaPv(ptr %add.ptr)
+  call void @_ZdaPv(ptr %add.ptr)
```
Comment (nunoplopes) [inline tests/alive-tv/bugs/pr22727.srctgt.ll:42]: revert

File: ir/instr.h [hunk 6]
```diff
@@ -1048,13 +1058,17 @@ class Memcpy final : public MemInstr {
 class Memcmp final : public MemInstr {
   Value *ptr1, *ptr2, *num;
   bool is_bcmp;
+  bool is_tailcall;
+
 public:
   Memcmp(Type &type, std::string &&name, Value &ptr1, Value &ptr2, Value &num,
-         bool is_bcmp): MemInstr(type, std::move(name)), ptr1(&ptr1),
-                        ptr2(&ptr2), num(&num), is_bcmp(is_bcmp) {}
+         bool is_bcmp, bool is_tailcall): MemInstr(type, std::move(name)), ptr1(&ptr1),
```
Comment (nunoplopes) [inline ir/instr.h:1065]: this is over 80 chars

File: ir/state.h [hunk 7]
```diff
@@ -171,6 +171,9 @@ class State {
   // The value of a 'returned' input
   std::optional<StateValue> returned_input;
 
+  // Used to keep track of the current instruction
+  const Value *current_value = nullptr;
```
Comment (nunoplopes) [inline ir/state.h:175]: this doesn't seem needed. Instructions can pass *this to check_tailcall.

File: ir/instr.cpp [hunk 8]
```diff
@@ -2313,6 +2313,35 @@ void FnCall::print(ostream &os) const {
...
 }
 
+static void check_tailcall(State &s) {
+  auto *call = static_cast<const Instr *>(s.getCurrentValue());
+  if (!call)
+    return;
+
+  auto it = s.getFn().bbOf(*call).instrs().begin();
+  bool found = false;
+  for (auto e = s.getFn().bbOf(*call).instrs().end(); it != e; ++it) {
```
Comment (nunoplopes) [inline ir/instr.cpp:2323]: pls don't call bbOf() twice. you can have a single loop to fetch the instruction after the call.

**Comment by nunoplopes** [comment]
I think overall this is the right direction, thanks!
Besides the comments I left, the printing is missing (it needs to print "tail " when appropriate).

File: tests/alive-tv/bugs/pr36437-bounded.srctgt.ll (hunk 3 above)
Comment (antoniofrighetto) [reply tests/alive-tv/bugs/pr36437-bounded.srctgt.ll:27]: We are not returning the value of the tail'd call site here, would be UB?

File: tests/alive-tv/bugs/pr18223.srctgt.ll (hunk 4 above)
Comment (antoniofrighetto) [reply tests/alive-tv/bugs/pr18223.srctgt.ll:77]: Same as above.

File: tests/alive-tv/bugs/pr22727.srctgt.ll (hunk 5 above)
Comment (antoniofrighetto) [reply tests/alive-tv/bugs/pr22727.srctgt.ll:42]: Same as above.

File: ir/instr.cpp (hunk 8 above)
Comment (antoniofrighetto) [reply ir/instr.cpp]: Definitely, thanks.

**Comment by antoniofrighetto** [comment]
Addressed reviews, added printing tail where appropriate! Current test failures:
```
FAIL: Alive2 :: alive-tv/calls/fn-ptr-taken.srctgt.ll (299 of 1253)
FAIL: Alive2 :: alive-tv/non-ub-exploitation/fncall-memory-havoc.srctgt.ll (724 of 1253)
```

**Comment by nunoplopes** [comment]
> Addressed reviews, added printing tail where appropriate! Current test failures:
> 
> ```
> FAIL: Alive2 :: alive-tv/calls/fn-ptr-taken.srctgt.ll (299 of 1253)
> FAIL: Alive2 :: alive-tv/non-ub-exploitation/fncall-memory-havoc.srctgt.ll (724 of 1253)
> ```

why are these failing? Can't merge the PR without these fixed somehow.

**Comment by antoniofrighetto** [comment]
> > Addressed reviews, added printing tail where appropriate! Current test failures:
> > ```
> > FAIL: Alive2 :: alive-tv/calls/fn-ptr-taken.srctgt.ll (299 of 1253)
> > FAIL: Alive2 :: alive-tv/non-ub-exploitation/fncall-memory-havoc.srctgt.ll (724 of 1253)
> > ```
> 
> why are these failing? Can't merge the PR without these fixed somehow.

As mentioned above, the failures are happening as part of the missing tgt function (in the tests) being optimized by TailCallElim in:
https://github.com/AliveToolkit/alive2/blob/5c1d5bc32d5e435560ce7b7e096d001c0f17affb/tools/alive-tv.cpp#L180

Thus, the tail attribute gets set for the two calls in the target function (`fn-ptr-taken.srctgt.ll`). When we happen to enter `check_tailcall` for the tgt, we add UB and the validation fails. This can be fixed by providing the tgt function for these two tests, yet the underlying issue is still there. Maybe we could bail out from `check_tailcall` when examining the target if the source has no tail calls. Do you have any suggestion?

**Comment by nunoplopes** [comment]
I may have to revert this PR if the comments in #900 are not fixed soon as I'm seeing quite a few test failures in the LLVM test suite because of this commit.

**Pull request merged**